exo/inference/pytorch/model/hf.py

```diff
@@ -0,0 +1,107 @@
+import torch
+import torch.nn as nn
+import numpy as np
+
+from transformers import AutoModelForCausalLM, LlamaConfig, DynamicCache, Cache
+from exo.inference.shard import Shard
+from exo.helpers import DEBUG
+from typing import Tuple
+
+from .utils import sample_logits
+
+class ShardedHuggingFaceModel(torch.nn.Module):
+    def __init__(self, shard: Shard):
+        super(ShardedHuggingFaceModel, self).__init__()
+
+        if DEBUG >= 2:
+            print(f"\nShardedHuggingFaceModel init with shard {shard}")
+
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        self.shard = shard
+
+        
+
+        # Load the model
+        self.full_model = AutoModelForCausalLM.from_pretrained(
```

Comment:
Will this download the entire model?
We have code to selectively download the model from HuggingFace so you don't have to download all layers on every device: `exo/download`

Reply:
Also this won't work with our download progress code. We show in the TUI what the download progress of the model is.